### ISSUE TITLE
Offline plugin doesnt work with Logstash 6.2.4

### ISSUE THREAD

**Issue by ditlef9** [elastic/logstash#9576]
Package used logstash-6.2.4.rpm.

Linux
```
cat /etc/*-release
LSB_VERSION=base-4.0-amd64:base-4.0-noarch:core-4.0-amd64:core-4.0-noarch
Red Hat Enterprise Linux Server release 6.9 (Santiago)
Red Hat Enterprise Linux Server release 6.9 (Santiago)

```

Java:
```
openjdk version "1.8.0_161"
OpenJDK Runtime Environment (build 1.8.0_161-b14)
OpenJDK 64-Bit Server VM (build 25.161-b14, mixed mode)
```

Online server :
`    bin/logstash-plugin prepare-offline-pack logstash-filter-dissect`

Offline server :
```
unzip logstash-offline-plugins-6.2.4.zip
sudo bin/logstash-plugin install --local /usr/share/logstash/logstash/logstash-filter-dissect-1.1.4.gem 

ERROR: Something went wrong when installing /usr/share/logstash/logstash/logstash-filter-dissect-1.1.4.gem, message: Failed to open TCP connection to artifacts.elastic.co:443 (execution expired)

```

**Comment by Bruce17** [comment]
Have the same error here to with logstash v6.2.3

**Comment by jsvd** [comment]
Have you tried to install the pack as described in [the offline pack docs](https://www.elastic.co/guide/en/logstash/current/offline-plugins.html#installing-offline-packs). Instead of unzipping, do:

```
bin/logstash-plugin install file:///path/to/logstash-offline-plugins-6.2.4.zip
```

**Comment by Bruce17** [comment]
I found a dirty workaround to get it working.

Edit the file `lib/pluginmanager/utils/http_client.rb` and replace `proxy_url = ENV["https_proxy"] || ENV["HTTPS_PROXY"] || ""` by `proxy_url = ENV["http_proxy"] || ENV["HTTP_PROXY"] || ""`.

Setting the `HTTPS_PROXY` caused some other errors in my case so I was dorced to used the `HTTP_PROXY`.

There should by some logic to use `HTTPS_PROXY` and as fallback `HTTP_PROXY` IMHO e.g.:

```ruby
    def self.start(uri)
      uri = URI(uri)
      proxy_url = ENV["https_proxy"] || ENV["HTTPS_PROXY"] || ENV["http_proxy"] || ENV["HTTP_PROXY"] || ""
      proxy_uri = URI(proxy_url)

      Net::HTTP.start(uri.host, uri.port, proxy_uri.host, proxy_uri.port, proxy_uri.user, proxy_uri.password, http_options(uri)) { |http| yield http }
    end
```

**Comment by Bruce17** [comment]
@jsvd Tried all cases. The program still tries to fetch the plugins's http head from the remote repository to check if the plugin exists.

**Comment by jsvd** [comment]
This is strange, I just tried locally and worked correctly:

```
/tmp/test % ls
offline/ online/
/tmp/test % cd offline
/tmp/test/offline % tar -zxf ~/Downloads/logstash-6.2.4.tar.gz
/tmp/test/offline % cd ../online
/tmp/test/online % tar -zxf ~/Downloads/logstash-6.2.4.tar.gz
/tmp/test/online % cd logstash-6.2.4
/tmp/test/online/logstash-6.2.4 % bin/logstash-plugin update logstash-filter-dissect
Updating logstash-filter-dissect
Updated logstash-filter-dissect 1.1.4 to 1.2.0
/tmp/test/online/logstash-6.2.4 % bin/logstash-plugin prepare-offline-pack logstash-filter-dissect
Offline package created at: /tmp/test/online/logstash-6.2.4/logstash-offline-plugins-6.2.4.zip

You can install it with this command `bin/logstash-plugin install file:///tmp/test/online/logstash-6.2.4/logstash-offline-plugins-6.2.4.zip`
/tmp/test/online/logstash-6.2.4 % cd ../../offline/logstash-6.2.4
/tmp/test/offline/logstash-6.2.4 % ping www.google.com
PING www.google.com (172.217.19.228): 56 data bytes
64 bytes from 172.217.19.228: icmp_seq=0 ttl=51 time=431.762 ms
^C
--- www.google.com ping statistics ---
1 packets transmitted, 1 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 431.762/431.762/431.762/0.000 ms
/tmp/test/offline/logstash-6.2.4 % ping www.google.com
ping: cannot resolve www.google.com: Unknown host
/tmp/test/offline/logstash-6.2.4 % bin/logstash-plugin install file:///tmp/test/online/logstash-6.2.4/logstash-offline-plugins-6.2.4.zip
Installing file: /tmp/test/online/logstash-6.2.4/logstash-offline-plugins-6.2.4.zip
Install successful
/tmp/test/offline/logstash-6.2.4 % bin/logstash-plugin list --verbose dissect
logstash-filter-dissect (1.2.0)
/tmp/test/offline/logstash-6.2.4 % ping www.google.com
ping: cannot resolve www.google.com: Unknown host
```

**Comment by Bruce17** [comment]
@jsvd Try it on an environment with no internet connection at all and only a proxy.

**Comment by jsvd** [comment]
I'm not sure what you mean there by "no internet connection at all", before installing the pack I disabled the network card (shown by not being able to contact Google).

As for the proxy, does it mean that having the proxy is necessary to see the failure? Does unsetting the proxy environment variables make it work?

**Comment by Bruce17** [comment]
I'm here behind a corporate proxy, so this is the only way getting out to the internet.

Unsetting the proxy env will throw an exception on `install`, but not on `prepare-offline-pack`.

Setting the proxy env caused the exact error above on `install` and `prepare-offline-pack`: `[...] Failed to open TCP connection to artifacts.elastic.co:443 (execution expired)`.

**Comment by jsvd** [comment]
> Unsetting the proxy env will throw an exception on install, but not on prepare-offline-pack.

so without proxy nor internet connection you're seeing an error on installing the zip..this is strange, can you show the full command + error for this scenario?

As for the prepare-offline-pack + proxy environment variable that seems like a bug :(

**Comment by jsvd** [comment]
@Bruce17 it all seems to work. I set up a VM with a host only connection to the host machine that is running squid. So the VM only had access to the internet through the proxy:

```
vagrant@packer-virtualbox-iso-1530453096:~$ http_proxy=http://192.168.33.1:3128 curl -I example.org
HTTP/1.1 200 OK
Cache-Control: max-age=604800
Content-Type: text/html
Date: Wed, 01 Aug 2018 14:46:54 GMT
ETag: "1541025663+ident"
Expires: Wed, 08 Aug 2018 14:46:54 GMT
Last-Modified: Fri, 09 Aug 2013 23:54:35 GMT
Server: ECS (dca/24F9)
Vary: Accept-Encoding
X-Cache: HIT
Content-Length: 1270
Age: 2734
X-Cache: HIT from 7c491948a3d6
X-Cache-Lookup: HIT from 7c491948a3d6:3128
Via: 1.1 7c491948a3d6 (squid/3.5.12)
Connection: keep-alive

vagrant@packer-virtualbox-iso-1530453096:~$ curl -I example.org
curl: (6) Could not resolve host: example.org
```

Then I downloaded logstash (through the proxy), set both https_proxy and http_proxy to the proxy, updated the dissect filter and created a pack:

```
vagrant@packer-virtualbox-iso-1530453096:~$ https_proxy=http://192.168.33.1:3128 http_proxy=http://192.168.33.1:3128 wget https://artifacts.elastic.co/downloads/logstash/logstash-6.2.4.tar.gz
--2018-08-01 15:33:17--  https://artifacts.elastic.co/downloads/logstash/logstash-6.2.4.tar.gz
Connecting to 192.168.33.1:3128... connected.
Proxy request sent, awaiting response... 200 OK
Length: 147224408 (140M) [binary/octet-stream]
Saving to: ‘logstash-6.2.4.tar.gz’

logstash-6.2.4.tar.gz           100%[======================================================>] 140.40M  3.76MB/s    in 43s     

2018-08-01 15:34:00 (3.28 MB/s) - ‘logstash-6.2.4.tar.gz’ saved [147224408/147224408]

vagrant@packer-virtualbox-iso-1530453096:~$ mkdir package
vagrant@packer-virtualbox-iso-1530453096:~$ cd package/
vagrant@packer-virtualbox-iso-1530453096:~/package$ tar -zxf ../logstash-6.2.4.tar.gz 
vagrant@packer-virtualbox-iso-1530453096:~/package$ cd logstash-6.2.4/
vagrant@packer-virtualbox-iso-1530453096:~/package/logstash-6.2.4$ export https_proxy=http://192.168.33.1:3128 && export http_proxy=http://192.168.33.1:3128
vagrant@packer-virtualbox-iso-1530453096:~/package/logstash-6.2.4$ bin/logstash-plugin update logstash-filter-dissect
Updating logstash-filter-dissect
Updated logstash-filter-dissect 1.1.4 to 1.2.0
vagrant@packer-virtualbox-iso-1530453096:~/package/logstash-6.2.4$ bin/logstash-plugin prepare-offline-pack logstash-filter-dissect
Offline package created at: /home/vagrant/package/logstash-6.2.4/logstash-offline-plugins-6.2.4.zip

You can install it with this command `bin/logstash-plugin install file:///home/vagrant/package/logstash-6.2.4/logstash-offline-plugins-6.2.4.zip`
```

Now, still with the environment variables for the proxies defined, I set up another logstash directory and installed the pack:

```
vagrant@packer-virtualbox-iso-1530453096:~/package/logstash-6.2.4$ cd ..
vagrant@packer-virtualbox-iso-1530453096:~/package$ cd ..
vagrant@packer-virtualbox-iso-1530453096:~$ mkdir install
vagrant@packer-virtualbox-iso-1530453096:~$ cd install
vagrant@packer-virtualbox-iso-1530453096:~/install$ tar -zxf ../logstash-6.2.4.tar.gz 
vagrant@packer-virtualbox-iso-1530453096:~/install$ cd logstash-6.2.4/
vagrant@packer-virtualbox-iso-1530453096:~/install/logstash-6.2.4$ curl -I example.org
HTTP/1.1 200 OK
Cache-Control: max-age=604800
Content-Type: text/html
Date: Wed, 01 Aug 2018 14:46:54 GMT
ETag: "1541025663+ident"
Expires: Wed, 08 Aug 2018 14:46:54 GMT
Last-Modified: Fri, 09 Aug 2013 23:54:35 GMT
Server: ECS (dca/24F9)
Vary: Accept-Encoding
X-Cache: HIT
Content-Length: 1270
Age: 3034
X-Cache: HIT from 7c491948a3d6
X-Cache-Lookup: HIT from 7c491948a3d6:3128
Via: 1.1 7c491948a3d6 (squid/3.5.12)
Connection: keep-alive

vagrant@packer-virtualbox-iso-1530453096:~/install/logstash-6.2.4$ http_proxy= https_proxy= curl -I example.org
curl: (6) Could not resolve host: example.org
vagrant@packer-virtualbox-iso-1530453096:~/install/logstash-6.2.4$ bin/logstash-plugin install file:///home/vagrant/package/logstash-6.2.4/logstash-offline-plugins-6.2.4.zip
Installing file: /home/vagrant/package/logstash-6.2.4/logstash-offline-plugins-6.2.4.zip
Install successful
```

**Comment by Bruce17** [comment]
Ok I doubled checked everything again with v6.3.2 and v6.2.3. Strange, today it's working. Seems there were network issues yesterday. Sorry about that.

Results:

<details>
<summary>Using Logstash <b>v6.3.2</b></summary>
<p>
1. Logstash already installed via package manager


2. Check internet connection:
```
$ curl -I www.google.de

curl: (7) Failed to connect to www.google.de port 80: Connection timed out
```

=> Internet access not possible


3. Check internet connection using proxy:
```
$ curl -I www.google.de --proxy dummy-proxy.local:3128

HTTP/1.1 200 OK
Via: 1.1 SEDCIPRX0330P15
Connection: close
Proxy-Connection: close
Expires: -1
Date: Thu, 02 Aug 2018 05:55:11 GMT
Content-Type: text/html; charset=ISO-8859-1
Server: gws
Accept-Ranges: none
Cache-Control: private, max-age=0
P3P: CP="This is not a P3P policy! See g.co/p3phelp for more info."
X-XSS-Protection: 1; mode=block
X-Frame-Options: SAMEORIGIN
Set-Cookie: 1P_JAR=2018-08-02-05; expires=Sat, 01-Sep-2018 05:55:11 GMT; path=/; domain=.google.de
Set-Cookie: NID=135=[...]; expires=Fri, 01-Feb-2019 05:55:11 GMT; path=/; domain=.google.de; HttpOnly
Vary: Accept-Encoding
```

=> Internet access possible.


4. Install plugin "logstash-input-lumberjack"
```
$ http_proxy=http://dummy-proxy.local:3128 https_proxy=http://dummy-proxy.local:3128  bin/logstash-plugin install logstash-input-lumberjack

WARNING: A maven settings file already exist at /root/.m2/settings.xml, please review the content to make sure it include your proxies configuration.
Validating logstash-input-lumberjack
Installing logstash-input-lumberjack
Installation successful
```


5. Try to update plugin "logstash-filter-dissect"
```
$ http_proxy=http://dummy-proxy.local:3128 https_proxy=http://dummy-proxy.local:3128  bin/logstash-plugin update logstash-filter-dissect

WARNING: A maven settings file already exist at /root/.m2/settings.xml, please review the content to make sure it include your proxies configuration.
Updating logstash-filter-dissect
No plugin updated
```


6. Try to do an offline pack
```
$ http_proxy=http://dummy-proxy.local:3128 https_proxy=http://dummy-proxy.local:3128  bin/logstash-plugin prepare-offline-pack logstash-input-lumberjack

WARNING: A maven settings file already exist at /root/.m2/settings.xml, please review the content to make sure it include your proxies configuration.
Offline package created at: /usr/share/logstash/logstash-offline-plugins-6.3.2.zip

You can install it with this command `bin/logstash-plugin install file:///usr/share/logstash/logstash-offline-plugins-6.3.2.zip`
```


7. Try to do an offline install
```
$ bin/logstash-plugin install file:///usr/share/logstash/logstash-offline-plugins-6.3.2.zip

Installing file: /usr/share/logstash/logstash-offline-plugins-6.3.2.zip
Install successful
```


8. Try to install a plugin with "--local"
```
$ bin/logstash-plugin install --local /path/to/logstash/local/plugins/logstash-input-lumberjack-3.1.5.gem

ERROR: Something went wrong when installing /path/to/logstash/local/plugins/logstash-input-lumberjack-3.1.5.gem, message: Failed to open TCP connection to artifacts.elastic.co:443 (execution expired)
```

9. Try to install a plugin with "--local" with proxy
```
$ http_proxy=http://dummy-proxy.local:3128 https_proxy=http://dummy-proxy.local:3128 bin/logstash-plugin install --local /path/to/logstash/local/plugins/logstash-input-lumberjack-3.1.5.gem

Validating /path/to/logstash/local/plugins/logstash-input-lumberjack-3.1.5.gem
Installing logstash-input-lumberjack
Installation successful
```

</p>
</details>

<details>
<summary>Using Logstash <b>v6.2.3</b></summary>
<p>
1. Logstash already installed via package manager


2. Check internet connection:
```
$ curl -I www.google.de

curl: (7) Failed to connect to www.google.de port 80: Connection timed out
```

=> Internet access not possible


3. Check internet connection using proxy:
```
$ curl -I www.google.de --proxy dummy-proxy.local:3128

HTTP/1.1 200 OK
Via: 1.1 SEDCIPRX0330P15
Connection: close
Proxy-Connection: close
Expires: -1
Date: Thu, 02 Aug 2018 06:15:12 GMT
Content-Type: text/html; charset=ISO-8859-1
Server: gws
Accept-Ranges: none
Cache-Control: private, max-age=0
P3P: CP="This is not a P3P policy! See g.co/p3phelp for more info."
X-XSS-Protection: 1; mode=block
X-Frame-Options: SAMEORIGIN
Set-Cookie: 1P_JAR=2018-08-02-06; expires=Sat, 01-Sep-2018 06:15:12 GMT; path=/; domain=.google.de
Set-Cookie: NID=135=[...]; expires=Fri, 01-Feb-2019 06:15:12 GMT; path=/; domain=.google.de; HttpOnly
Vary: Accept-Encoding
```

=> Internet access possible.


4. Install plugin "logstash-input-lumberjack"
```
$ http_proxy=http://dummy-proxy.local:3128 https_proxy=http://dummy-proxy.local:3128  bin/logstash-plugin install logstash-input-lumberjack

Validating logstash-input-lumberjack
Installing logstash-input-lumberjack
Installation successful
```


5. Try to update plugin "logstash-filter-dissect"
```
$ http_proxy=http://dummy-proxy.local:3128 https_proxy=http://dummy-proxy.local:3128  bin/logstash-plugin update logstash-filter-dissect

Updating logstash-filter-dissect
Updated logstash-filter-dissect 1.1.4 to 1.2.0
```


6. Try to do an offline pack
```
$ http_proxy=http://dummy-proxy.local:3128 https_proxy=http://dummy-proxy.local:3128  bin/logstash-plugin prepare-offline-pack logstash-input-lumberjack

Offline package created at: /usr/share/logstash/logstash-offline-plugins-6.2.3.zip

You can install it with this command `bin/logstash-plugin install file:///usr/share/logstash/logstash-offline-plugins-6.2.3.zip`
```


7. Try to do an offline install
```
$ bin/logstash-plugin install file:///usr/share/logstash/logstash-offline-plugins-6.3.2.zip

Installing file: /usr/share/logstash/logstash-offline-plugins-6.2.3.zip
Install successful
```


8. Try to install a plugin with "--local"
```
$ bin/logstash-plugin install --local /path/to/logstash/local/plugins/logstash-input-lumberjack-3.1.5.gem

ERROR: Something went wrong when installing /path/to/logstash/local/plugins/logstash-input-lumberjack-3.1.5.gem, message: Failed to open TCP connection to artifacts.elastic.co:443 (execution expired)
```


9. Try to install a plugin with "--local" with proxy
```
$ http_proxy=http://dummy-proxy.local:3128 https_proxy=http://dummy-proxy.local:3128 bin/logstash-plugin install --local /path/to/logstash/local/plugins/logstash-input-lumberjack-3.1.5.gem

Validating /path/to/logstash/local/plugins/logstash-input-lumberjack-3.1.5.gem
Installing logstash-input-lumberjack
Installation successful
```

</p>
</details>

**But**: I still have the issue with  `bin/logstash-plugin install --local [...]`. The documentation says: "install --local: force local-only plugin installation. see bin/logstash-plugin package|unpack (default: false)".

And:
* "pack: Package currently installed plugins, Deprecated: Please use prepare-offline-pack instead"
* "unpack: Unpack packaged plugins, Deprecated: Please use prepare-offline-pack instead"

So to sum it up: "install --local" should work like "prepare-offline-pack", but I have the issues you can see in the logs above.

**Comment by pathakshivam** [comment]
Well, the same problem occurs for me, I just cleared Auto proxy discovery and connected by machine to my personal network and it worked for me.

It is not the solution but a workaround if someone wants to get rid of this issue quickly.
